test/sslapitest.c

```diff
@@ -1967,12 +1967,12 @@ static const char *servalpn;
 static int alpn_select_cb (SSL *ssl, const unsigned char **out, unsigned char *outlen,
                     const unsigned char *in, unsigned int inlen, void *arg)
 {
-    unsigned int i, protlen = 0;
+    unsigned int protlen = 0;
     const unsigned char *prot;
 
-    for (i = 0, prot = in; i < inlen; i += protlen, prot += protlen) {
-        protlen = *(prot++);
-        if (inlen - i < protlen)
+    for (prot = in; prot < in + inlen; prot += protlen) {
+        protlen = *prot++;
+        if (in + inlen - prot < protlen)
             return SSL_TLSEXT_ERR_NOACK;
 
         if (protlen == strlen(servalpn)
```

Comment:
Why isn't this strlen for "goodalpn"? I mean since following memcmp is looking for "goodalpn" then one can as well compare to length of "goodalpn"...

Reply:
I didn't go through all cases in this file so I am not sure if this is intended here. One reason of not strlen for `"goodalpn"` is that the global `servalpn` would be changed to control the result of this `alpn_select_cb` callback, so it seems the `memcmp` should be changed to compare against `servalpn` instead of `"goodalpn"`...

Reply:
Yes, I think that was the intention of this code

Reply:
> Yes, I think that was the intention of this code

What was? To memcmp against servalpn and not "goodalpn"? Well, that would make sense too. Basically I failed to see the reason for why input to strlen and memcmp would be *different*. I assumed that intension was to catch "goodalpn", but it might as well be to catch servalpn whatever it might happen to be at the given moment...

Reply:
I meant the intention was to memcmp against servalpn.

Reply:
So it's kind of a bug. Do we fix it here or in separate request? I'd say it's not inappropriate to fix it right away, but separate commit might be appropriate...

Reply:
The fix could be very straightforward, so this should be fixed right away - in a separate commit in this PR...